core/src/HIP/Kokkos_HIP_Atomic.hpp

```diff
@@ -287,210 +284,285 @@ namespace Kokkos {
-        *dest = return_val + val;
-        ::Kokkos::Impl::unlock_address_hip_space( (void*) dest );
-        done = 1;
-      }
+// WORKAROUND
+template <class T>
+KOKKOS_INLINE_FUNCTION T atomic_fetch_add(
+    volatile T *dest, typename std::enable_if<sizeof(T) != sizeof(int) &&
+                                                  sizeof(T) != sizeof(int64_t),
+                                              const T &>::type val)
+{
+  T return_val;
+  // Do we need to (like in CUDA) handle potential wavefront branching?
+  int done = 0;
+  // unsigned int active = KOKKOS_IMPL_CUDA_BALLOT(1);
+  // unsigned int done_active = 0;
+  // while (active!=done_active) {
+  if (!done)
+  {
+    bool locked = ::Kokkos::Impl::lock_address_hip_space((void *)dest);
+    if (locked)
+    {
+      return_val = *dest;
+      *dest = return_val + val;
+      ::Kokkos::Impl::unlock_address_hip_space((void *)dest);
+      done = 1;
     }
-    //done_active = KOKKOS_IMPL_CUDA_BALLOT(done);
-  //}
-    return return_val;
   }
+  // done_active = KOKKOS_IMPL_CUDA_BALLOT(done);
+  //}
+  return return_val;
+}
 
-  template<class T>
-  KOKKOS_INLINE_FUNCTION
-  T atomic_fetch_sub(volatile T* dest, typename std::enable_if<sizeof(T) == sizeof(int),T>::type val) {
-    union U {
-      int i ;
-      T t ;
-      KOKKOS_INLINE_FUNCTION U() {};
-    } assume , oldval , newval ;
+*/
 
-    oldval.t = *dest ;
+KOKKOS_INLINE_FUNCTION
+int atomic_fetch_sub(volatile int *dest, int const &val)
+{
+  return atomicSub(const_cast<int *>(dest), val);
+}
 
-    do {
-      assume.i = oldval.i ;
-      newval.t = assume.t - val ;
-      oldval.i = Kokkos::atomic_compare_exchange( (int*)dest , assume.i , newval.i );
-    } while ( assume.i != oldval.i );
+KOKKOS_INLINE_FUNCTION
+unsigned int atomic_fetch_sub(volatile unsigned int *dest,
+                              unsigned int const &val)
+{
+  return atomicSub(const_cast<unsigned int *>(dest), val);
+}
 
-    return oldval.t ;
-  }
+KOKKOS_INLINE_FUNCTION
+int64_t atomic_fetch_sub(int64_t *dest, int64_t const &val)
+{
+  return static_cast<int64_t>(
+      atomicAdd(reinterpret_cast<unsigned long long *>(dest),
+                -reinterpret_cast<unsigned long long const &>(val)));
+}
 
-  template<class T>
-  KOKKOS_INLINE_FUNCTION
-  T atomic_fetch_sub(volatile T* dest, typename std::enable_if<sizeof(T) != sizeof(int) && sizeof(T) == sizeof(int64_t), const T&>::type val) {
-    union U {
-      int64_t i ;
-      T t ;
-      KOKKOS_INLINE_FUNCTION U() {};
-    } assume , oldval , newval ;
+KOKKOS_INLINE_FUNCTION
+char atomic_fetch_sub(volatile char *dest, const char &val)
+{
+  unsigned int oldval, newval, assume;
+  oldval = *(int *)dest;
+
+  do
+  {
+    assume = oldval;
+    newval = assume & 0x7fffff00 + ((assume & 0xff) - val) & 0xff;
+    oldval = hc::atomic_compare_exchange_unsigned((unsigned int *)dest, assume,
+                                                  newval);
+  } while (assume != oldval);
+
+  return oldval;
```

Comment:
Where does that come from?

Reply:
Here https://github.com/ORNL-CEES/kokkos/pull/10/files#diff-ed152c9ca6d8623c27b861a713236a76L319 I just uncommented the code